File: src/0847-Shortest-Path-Visiting-All-Nodes/0847.go
func shortestPathLength(graph [][]int) int {
    n := uint(len(graph))
    t, res := uint((1 << n) - 1), 0
    q := make([][]uint, n)
    mem := make([][]uint, n)
    for i := uint(0); i < n; i++ {
        mem[i] = make([]uint, 1 << n);
        q[i] = []uint{i, 1 << i}
    }
    for true {
        nq := len(q)
        for i := 0; i < nq; i++ {
            pre := q[0]
            q = q[1:]
            if pre[1] == t {
                return res
            }
            for _, j := range graph[pre[0]] {
                k, new_stat := uint(j), pre[1] | 1 << uint(j)
                if mem[k][new_stat] == 0 {
                    q = append(q, []uint{k, new_stat})
                    mem[k][new_stat] = 1
                }
            }
        }
        res++
    }
    return 0
}

File: src/0847-Shortest-Path-Visiting-All-Nodes/0847.js
var shortestPathLength = function(graph) {
    let n = graph.length;
    let t = (1 << n) - 1, res = 0, q = [];
    let mem = new Array(n);
    for (let i = 0; i < n; i++) {
        mem[i] = new Array(1 << n);
        mem[i].fill(0);
        q[i] = [i, 1 << i];
    }
    while (true) {
        let nq = q.length;
        for (let i = 0; i < nq; ++i) {
            let pre = q.shift();
            if (pre[1] == t) return res;
            for (let j of graph[pre[0]]) {
                let new_stat = pre[1] | 1 << j;
                if (!mem[j][new_stat]) {
                    q.push([j, new_stat]);
                    mem[j][new_stat] = 1;
                }
            }
        }
        res++;
    }
};